perfkitbenchmarker/providers/azure/azure_blob_storage.py

```diff
@@ -0,0 +1,119 @@
+# Copyright 2016 PerfKitBenchmarker Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+
+from perfkitbenchmarker import flags
+from perfkitbenchmarker import linux_packages
+from perfkitbenchmarker import object_storage_service
+from perfkitbenchmarker import providers
+from perfkitbenchmarker import vm_util
+
+flags.DEFINE_string('azure_lib_version', None,
+                    'Use a particular version of azure client lib, e.g.: 1.0.2')
+
+FLAGS = flags.FLAGS
+
+DEFAULT_AZURE_REGION = 'East US'
+AZURE_CREDENTIAL_LOCATION = '.azure'
+
+
+class AzureBlobStorageService(object_storage_service.ObjectStorageService):
+  """Interface to Azure Blob Storage."""
+
+  STORAGE_NAME = providers.AZURE
+
+  def PrepareService(self, location):
+    self.storage_account = 'pkb%s' % FLAGS.run_uri
+    vm_util.IssueCommand(
+        ['azure', 'storage', 'account', 'create',
+         '--type', 'ZRS',
+         '-l', location or DEFAULT_AZURE_REGION,
+         self.storage_account])
+
+    output, _, _ = vm_util.IssueCommand(
+        ['azure', 'storage', 'account',
+         'keys', 'list', self.storage_account])
+
+    key = re.findall(r'Primary:* (.+)', output)
+    self.azure_key = key[0]
+
+  def CleanupService(self):
+    vm_util.IssueCommand(
+        ['azure', 'storage', 'account', 'delete',
+         '-q', self.storage_account])
+
+  def MakeBucket(self, bucket):
+    vm_util.IssueCommand(
+        ['azure', 'storage', 'container',
+         'create', bucket,
+         '-a', self.storage_account,
+         '-k', self.azure_key])
+
+  def DeleteBucket(self, bucket):
+    vm_util.IssueCommand(
+        ['azure', 'storage', 'container',
+         'delete', '-q', bucket,
+         '-a', self.storage_account,
+         '-k', self.azure_key])
+
+  def EmptyBucket(self, bucket):
+    # Emptying buckets on Azure is hard. We pass for now - this will
+    # increase our use of storage space, but should not affect the
+    # benchmark results.
+    pass
```

Comment:
I'm not sure I understand the question. The function is mostly a space-saving optimization (and therefore money-saving optimization) for long-running benchmarks. In the CLI-based benchmark, where we upload files and then download them many times, this is the function that empties the bucket after each download. The cleanup at the end of the benchmark is the `DeleteBucket` function above, which works even with objects in the bucket.

On GCS and S3, you can issue a single command to delete all objects in a bucket. On Azure, it appears that you have to list the contents of the bucket and then issue delete commands for every object individually, which is one reason I didn't want to implement this. In the current benchmark, we empty buckets on Azure using the API test script, but on GCS and S3 using the CLI tool, which feels very inconsistent to me.